### DISCUSSION TITLE
Has anyone made it to train a working FineTransformer?

### DISCUSSION THREAD

**Discussion by cyanbx** [lucidrains/audiolm-pytorch#116] [Q&A]
I have successfully trained a CoraseTransformer which can generate intelligable speech from semantic tokens. But although the FineTransformer seems to have a similar architecture and training pipeline, I can't reconstruct high quality audio with it. It's output audio doesn't seem to be better than those from only coarse tokens, and the spectrogram graphs seem to be more blurred. Is there any advice on improving the training or inference process?

My FineTransformer hparams:
```
fine_transformer = FineTransformer(
    num_coarse_quantizers = 3,
    num_fine_quantizers = 5,
    codebook_size = 1024,
    dim = 512,
    depth = 6
)
```

Here are the comparison of spectrograms.
Reconstructed from COARSE TOKEN:
![coarse0](https://user-images.githubusercontent.com/50834547/222155582-f8548b24-de98-4c65-abe1-cd388c4ebc89.png)
Reconstructed from COARSE TOKEN and Inferenced FINE TOKEN:
![fine0](https://user-images.githubusercontent.com/50834547/222155826-5a3a8dc6-a3a6-4940-bcb0-5d177e31775d.png)
Reconstructed from Ground Truth TOKEN:
![compressed0](https://user-images.githubusercontent.com/50834547/222155810-cc400062-c5cf-46ba-b3a1-471040fc9fdb.png)

**Comment by lucidrains** [comment]
@cyanbx thanks for this! i'll review the fine transformer code today

i'm sure it is something minor

**Reply by lucidrains** [reply]
@cyanbx judging from your comments asking about soundstream, it could also be that your soundstream is undertrained, so the fine quantize ids don't really do that much

**Reply by cyanbx** [reply]
Thanks for your fast reply. But actually I didn't use soundstream provided in this repo. I alternate it with the pretrained encodec (https://github.com/facebookresearch/encodec), which should have high audio quality。

**Reply by lucidrains** [reply]
@cyanbx if you have time, could you try training a very small soundstream, say with only 4 quantizers, and if the fine transformer still does not perform well, we can rule in an underlying issue with the transformer logic

**Reply by lucidrains** [reply]
@cyanbx ohh got it, then there must be an issue, thanks, will look into it!

**Reply by lucidrains** [reply]
@cyanbx last question, are you sure encodec has the quantizer dimension in the right place for the transformers in this repository to pick up on? or did you write a wrapper around encodec and make sure all the functions return correct tensors?

**Reply by cyanbx** [reply]
I think encodec has the right quantizer dimention for the transformer, since I have trained coarse transformer with it, and get intelligible speeches.

**Reply by lucidrains** [reply]
@cyanbx ok good to know, i'll look over the code tonight! thank you!

**Reply by cyanbx** [reply]
Again, thanks for your kind help.

I'm now trying to train a fine transformer with 4 codebooks, 1 for coarse and 3 for fine, and see whether it will show stronger contrast.

I'm not quite sure about so forgive me if I'm wrong, but it seems that the transformer has inconsistent behavior during training and inferencing. Do you think attend to the generated fine tokens (which are not done during training) will affect the performance?

**Reply by lucidrains** [reply]
@cyanbx hmm can you point me to where you think fine tokens are not attended to?

i see it being passed into the transformers [here](https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/audiolm_pytorch.py#L1511)

**Reply by cyanbx** [reply]
Ohh, I thought it was masked by `self_attn_mask` [here](https://github.com/lucidrains/audiolm-pytorch/blob/main/audiolm_pytorch/audiolm_pytorch.py#L1502), but it seems I misunderstood it. Sorry about that.

**Reply by lucidrains** [reply]
@cyanbx i couldn't find an issue, but i made a simplification to the fine transformer in [0.19.0](https://github.com/lucidrains/audiolm-pytorch/commit/65495ad5b060bfc90d056d4a161d2df5060eea6a)

**Reply by lucidrains** [reply]
@cyanbx are you still training this? will be implementing a specifically tailored relative positional encoding for the fine transformer this morning (so that the coarse and fine sequences have knowledge of each other's relative distances)

**Reply by cyanbx** [reply]
@lucidrains I have trained a 1coarse3fine version, which works well. But then the performance degrades when I train a 3coarse13fine version. It seems that longer sequence may cause performance degradation. Hopefully the new positional encoding helps.

**Reply by lucidrains** [reply]
@cyanbx oh good to know about the 1coarse3fine! i think i can fix this for you in the next hour!

**Reply by lucidrains** [reply]
@cyanbx ok, should be easier to train! https://github.com/lucidrains/audiolm-pytorch/commit/af903342a19b88f5bce8479bfab8d8aa862f86b9

**Reply by zhvng** [reply]
@cyanbx  did this end up working for you? just wondering

**Reply by cyanbx** [reply]
@zhvng Not actually. The loss is still high during training.

**Reply by lucidrains** [reply]
@cyanbx ah doh, thought it was a shoo-in

i'm moving onto other work, but could you possibly identify at which length for coarse and fine it starts to train not as well, for when i return and can put more time into this?

in the worst case, i can bring in the design from a [lesser known paper](https://github.com/lucidrains/RQ-Transformer), but it will take a time to adapt it for audio lm

**Reply by lucidrains** [reply]
actually, maybe i can bet on one last simple intervention that could do the trick

**Reply by lucidrains** [reply]
@cyanbx ok try 0.23.0, and if that doesn't work, still have 2 more tricks up my sleeves before we go the RQ-transformer route. will circle back near end of the month!

**Reply by cyanbx** [reply]
@lucidrains Thanks for your great effort. I'm actually working on some other experiments related to this project, but I will definitely try it sometime later.

**Reply by lucidrains** [reply]
@cyanbx ok cool np, yea, tidy up a few experiments that you can share online (perhaps with wandb.ai) for when i return and can deal a few more blows to this problem space

**Reply by lzl1456** [reply]
Great work by this author ，copy his  work  @cyanbx，I have also successfully trained a CoraseTransformer which can generate intelligable speech from semantic tokens.  but Compared with the original speech, The content of the generated speech is consistent for a period of time (probably less than 3 seconds), but the latter part of the content is more like random generation. Is this related to the training data and duration?

**Reply by eonglints** [reply]
Just to add that another option would be to go for the non-autoregressive approach that the VALL-E authors use. This might also help with the long sequence lengths. However, I realise it's diverging from AudioLM quite a bit.

> "The NAR model has a similar architecture to the AR model, except that it contains eight separate acoustic embedding layers. In each training step, we randomly sample a training stage _i_ ∈ [2, 8]. The model is trained to maximize the acoustic tokens from the _i_-th quantizer codebook."
(from the [VALL-E paper](https://arxiv.org/abs/2301.02111), section 4.2.2)

**Reply by lucidrains** [reply]
@eonglints oh yes, that's a good idea

will implement if on my return i see researchers still having issues!

**Reply by eonglints** [reply]
Look

> @eonglints oh yes, that's a good idea
> 
> will implement if on my return i see researchers still having issues!

Looks like we may as well jump over to SoundStorm for this, thanks for starting [the implementation](https://github.com/lucidrains/soundstorm-pytorch), I'll try and start testing next week.

**Reply by lucidrains** [reply]
@eonglints that could have been your paper!

**Reply by lucidrains** [reply]
@eonglints are you still trying to get into a phd program atm? what are you up to?

**Reply by eonglints** [reply]
I just started at Stability AI a few weeks ago :)

Also still doing my PhD at the University of Edinburgh.

When VALL-E came out I was really pleased to to see TTS with audio language models working but yeah, tiny bit frustrated that we hadn't quite got there first on this repo! Just got very busy with other things. Back in the thick of it now though.

**Reply by lucidrains** [reply]
congrats! they are lucky to have you!

**Reply by eonglints** [reply]
Aw, thanks. It's going to be fun.

**Comment by Liujingxiu23** [comment]
@cyanbx  Hi， I am now also using codecs of Encoder(facebook  pretrained version) to train a coasre2fine model. My problem is training in single-gpu works，but when multi-gpus, I can not start the train.   The details is in https://github.com/lucidrains/audiolm-pytorch/discussions/128
Could you please help me figure the problem? or Could you share your training code? I mean the dataset and  dataloader part.
Thank you !

**Reply by cyanbx** [reply]
Sorry but I have only trained on a single-gpu with this architecture.

**Reply by Liujingxiu23** [reply]
Thanks for your reply.
Are you still working on this model , and have ever get good results?

I tried 16-codec book data from Encodec model of facebook, 6 as coasre and 10 as fine. And have trained on one GPU for about one week. But the synthesized result are even worse then I just let all values of 10 fine codebooks as zero.

**Reply by lzl1456** [reply]
Can't you use his original code for more cards?

**Reply by Liujingxiu23** [reply]
I can not run the original code since I do not have a good soundstream model. I have to revised part of the code to support my trainning of my own data,  details  in https://github.com/lucidrains/audiolm-pytorch/discussions/128

**Reply by lzl1456** [reply]
oh,yes,,,I seem to have this problem too，encodec doesn't seem to load into accelerate，if you have solved this part, hope to see it together！

**Reply by LWprogramming** [reply]
Can you add your Encodec code (i.e. how are you using the API)? Would be interesting if your approach is similar to mine in #135 (PR not reviewed/ merged at the moment but I think it's probably at least close to usable)

I haven't done anything with `accelerate` though, is this something very critical to getting things to work?

**Reply by lzl1456** [reply]
@LWprogramming sorry ，i tried merge code ,  I don't have permission to upload code，My code is like below，I imported directly
![image](https://user-images.githubusercontent.com/38452732/227830311-b80ca6f1-1666-4177-9c91-11685878017d.png)

![image](https://user-images.githubusercontent.com/38452732/227830246-d5e652e8-a71b-409b-8988-ef9cab35acf8.png)

I read your code ,：
 Is the shape here needs to be modified?
![image](https://user-images.githubusercontent.com/38452732/227829740-febc0ba4-c524-4b33-b68d-63c7be1de077.png)

![image](https://user-images.githubusercontent.com/38452732/227829795-8b29125d-bf43-4b2c-b74b-9431fc2befbf.png)

![image](https://user-images.githubusercontent.com/38452732/227829829-d14b7a08-1f84-4b6a-b88c-4347249a9527.png)

**Reply by lzl1456** [reply]
This does not need to be used here, because the dataload has already resample, and this will change the batch to 1
![image](https://user-images.githubusercontent.com/38452732/227834478-8301422e-7b07-42b3-91a3-415a600e30d3.png)

![image](https://user-images.githubusercontent.com/38452732/227834491-76b17c9f-f467-41b6-a29b-498a4fb781a8.png)

**Reply by LWprogramming** [reply]
Ah yes @lzl1456 I think you are correct

re: shape, I think this is the same issue as in #155

re: convert_audio, good call, I'll make a pr #157

**Comment by huangf79** [comment]
@cyanbx  Hi, could you share your training details of CoraseTransformer and some audio samples generated by it? I'm trying to train a CoraseTransformer on the 'dev-clean' set of LibriSpeech and I only get noise even i use a training sample as input.

**Reply by amitaie** [reply]
I'm trying for some time to train CoraseTransformer, but it doesn't work properly for me. try to train CoraseTransformer that uses only 1 quantizer, it might works and will help you understand what is going on.
